### DISCUSSION TITLE
Quick questions about using Hyper with HTTP/2

### DISCUSSION THREAD

**Discussion by OldMansKid** [hyperium/hyper#3442] [Q&A] [answered]
Hello community,
I notice that http1::SendRequest doesn't implement Clone while http2::SendRequest does. On the other hand, hyper::service::Service isn't required to implement Clone. So, I assume that in HTTP/2, a single Service's call() function will be called with requests from all HTTP/2 streams, and cloning an http2::SendRequest will open a new stream, am I right?

I also notice that Hyper doesn't expose any APIs to operate on HTTP/2 pseudo headers nor on frames like sending RST_STREAM, while the h2 crate gives more control over these things. Does it mean that I need to use h2 instead of SendRequest/Builder APIs if I have such needs? Thanks in advance.

**Comment by seanmonstar** [comment, marked as answer]
> I assume that in HTTP/2, a single Service's call() function will be called with requests from all HTTP/2 streams, and cloning an http2::SendRequest will open a new stream, am I right?

Not quite. In HTTP/2, sending any request will always result in a new stream. The reason the `http1::SendRequest` isn't `Clone` is because an HTTP/1 connection cannot send new requests until the previous one has finished. Whereas `http2::SendRequest` can be cloned to send multiple requests at the same time. (The `Clone` call doesn't create a new stream. It's the `send_request()` call.)

> HTTP/2 pseudo headers

Most of them are translated to `http` types, either pieces of the `Uri`, or the `StatusCode`.

> frames like sending RST_STREAM

If you need to deal with those frames directly, yes you'll need to look at the `h2` frame. Though, hyper will use those protocol-specific frames by translating from the level that hyper exposes.
